### ISSUE TITLE
Whitelisting only certain file-types and performing security checks on them - AWS s3

### ISSUE THREAD

**Issue by vincentcox** [ankitpokhrel/tus-php#310]
I don't know if it's already possible, but I only want to allow certain filetypes (like images only for example). It cannot be checked by only the extension, because attackers can just rename the file. It needs to be checked on mime-type, what's inside the file etc (similar to guessExtension() in Symfony). 

If this is outside the scope of this project, is there a way that I can do these checks myself when the file is completely uploaded to the server (and not yet uploaded to a Cloud provider)? I have a setup that I want to upload to AWS S3, can you give me a few cues on how to do that? Thanks!

**Comment by ankitpokhrel** [comment]
Hi @vincentcox you can use [events](https://github.com/ankitpokhrel/tus-php#events) for this purpose. A `tus-server.upload.complete` event is fired once the upload is complete and you can utilize this hook to add your validation logic before uploading the file to the cloud provider.

```php
$server->event()->addListener('tus-server.upload.complete', function (\TusPhp\Events\TusEvent $event) {
    $fileMeta = $event->getFile()->details();

    // Add validation logic.
    // Upload to cloud provider.
});
```

**Comment by vincentcox** [comment]
Hi @ankitpokhrel, would this become something like this in Symfony?

```php
<?php
namespace App\Controller;
use TusPhp\Tus\Server as TusServer;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\Routing\Annotation\Route;
use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;

class TusController extends AbstractController
{
    /**
     * Create tus server. Route matches /tus/, /tus and /tus/* endpoints.
     *
     * @Route("/tus/", name="tus_post")
     * @Route("/tus/{token?}", name="tus", requirements={"token"=".+"})
     *
     * @param TusService $tusService
     *
     * @return Response
     */
    public function server(TusServer $server)
    {
        $response = $server->serve();

        $server->event()->addListener('tus-server.upload.complete', function (\TusPhp\Events\TusEvent $event) {
            $fileMeta = $event->getFile()->details();

            // Add validation logic.
            // Upload to cloud provider.
        });

        return $response->send();
    }
}
```

I want to use this in combination with https://uppy.io/, so I don't know if this is the correct approach. I know this can be out of scope for this project, but if you already used something (similar) like this and want to give me some cues, I would highly appreciate it. Many thanks and stay safe!

**Comment by ankitpokhrel** [comment]
Hi @vincentcox, you first need to attach the listener before serving the request. Something like this should work:
```php
public function server(TusServer $server)
{
    $server->event()->addListener('tus-server.upload.complete', function (\TusPhp\Events\TusEvent $event) {
        $fileMeta = $event->getFile()->details();

        // Add validation logic.
        // Upload to cloud provider.
    });

    $response = $server->serve();

    return $response->send();
}
```
